File: iterableapi/src/main/java/com/iterable/iterableapi/ImpressionData.java
package com.iterable.iterableapi;

import androidx.annotation.RestrictTo;

import java.util.Date;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class ImpressionData {
    final String messageId;
    final boolean silentInbox;
    int displayCount = 0;
    float duration = 0.0f;

    Date impressionStarted = null;

    ImpressionData(String messageId, boolean silentInbox) {
        this.messageId = messageId;
        this.silentInbox = silentInbox;
    }

    void startImpression() {
        this.impressionStarted = new Date();
    }

    void endImpression() {
        if (this.impressionStarted != null) {
            this.displayCount += 1;
            this.duration += (float) (new Date().getTime() - this.impressionStarted.getTime()) / 1000;
            this.impressionStarted = null;
        }
    }
}

File: iterableapi/src/main/java/com/iterable/iterableapi/ImpressionData.kt
package com.iterable.iterableapi

import androidx.annotation.RestrictTo

import java.util.Date

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
internal class ImpressionData(
    val messageId: String,
    val silentInbox: Boolean
) {
    var displayCount = 0
    var duration = 0.0f

    internal var impressionStarted: Date? = null

    fun startImpression() {
        this.impressionStarted = Date()
    }

    fun endImpression() {
        //increment count and add to duration if impression has been started
        impressionStarted?.let { started ->
            this.displayCount += 1
            this.duration += (Date().time - started.time).toFloat() / 1000
            this.impressionStarted = null
        }
    }
}
